### PR TITLE
feat(os): improve macos & linux support

### PR DESCRIPTION
**Pull request by noelmcloughlin**
This PR has some improvements and fixes

1.  Set and use `rootuser` and `rootgroup` per template formula (macos/bsd)
2.  Clarify GOPATH because its important - README special note
3.  Improve linux alternatives  yaml handling and SLS updates
4.  Fix some issues with `file/default/golang.sh.jinja`

### CODE REVIEW

**Comment by myii** [comment]
@noelmcloughlin After the recent batch of `yamllint` PRs (see https://github.com/saltstack-formulas/template-formula/pull/159 as a starting point), the `template-formula` provides this functionality at the end of [`map.jinja`](https://github.com/saltstack-formulas/template-formula/blob/7dc422a54636be56774550f1f875e7c8ab4d5b65/template/map.jinja#L52-L56) instead.  That's why the `lint` stage (will be) failing here.

**Comment by noelmcloughlin** [comment]
Thanks @myii I fixed things up and verified on Ubuntu 18.04

```
          ID: golang-package-archive-install-file-directory
    Function: file.directory
        Name: /usr/local/go1.10.1.linux-amd64
      Result: True
     Comment: Directory /usr/local/go1.10.1.linux-amd64 updated
     Started: 17:50:43.812630
    Duration: 4258.498 ms
     Changes:   
              ----------
              mode:
                  0755
----------
          ID: golang-package-archive-install-file-directory
    Function: file.directory
        Name: /usr/local/golang/packages
      Result: True
     Comment: The directory /usr/local/golang/packages is in the correct state
     Started: 17:50:48.071336
    Duration: 0.89 ms
     Changes:   
----------
          ID: golang-package-archive-install-archive-extracted
    Function: archive.extracted
        Name: /usr/local/go1.10.1.linux-amd64
      Result: True
     Comment: All files in archive are already present
     Started: 17:50:48.072617
    Duration: 4367.464 ms
     Changes:   
----------
          ID: golang-package-archive-install-home-alternative-install
    Function: alternatives.install
        Name: golang-home
      Result: True
     Comment: Alternative /usr/local/go1.10.1.linux-amd64/go for golang-home is already registered
     Started: 17:50:52.440274
    Duration: 7.784 ms
     Changes:   
----------
          ID: golang-package-archive-install-godoc-alternative-install
    Function: alternatives.install
        Name: link-godoc
      Result: True
     Comment: Alternative /usr/local/go1.10.1.linux-amd64/go/bin/godoc for link-godoc is already registered
     Started: 17:50:52.448597
    Duration: 7.766 ms
     Changes:   
----------
          ID: golang-package-archive-install-gofmt-alternative-install
    Function: alternatives.install
        Name: link-gofmt
      Result: True
     Comment: Alternative /usr/local/go1.10.1.linux-amd64/go/bin/gofmt for link-gofmt is already registered
     Started: 17:50:52.456634
    Duration: 7.748 ms
     Changes:   
----------
          ID: golang-package-archive-install-go-alternative-install
    Function: alternatives.install
        Name: link-go
      Result: True
     Comment: Alternative /usr/local/go1.10.1.linux-amd64/go/bin/go for link-go is already registered
     Started: 17:50:52.464646
    Duration: 7.79 ms
     Changes:   
----------
          ID: golang-config-file-managed-environ_file
    Function: file.managed
        Name: /etc/default/golang.sh
      Result: True
     Comment: File /etc/default/golang.sh updated
     Started: 17:50:52.472695
    Duration: 32.949 ms
     Changes:   
              ----------
              diff:
                  --- 
                  +++ 
                  @@ -3,8 +3,13 @@
                   # Your changes may be overwritten.
                   ########################################################################
                   
                  -export GOROOT=/usr/local/go
                  -export GOPATH=/usr/local/golang/packages
                  -export GOBASE=/usr/local/go1.10.1.linux-amd64/go
                  -export PATH=$PATH:$GOROOT/bin:$GOBASE/bin
                  -export PATH=${PATH}:/usr/local/go1.10.1.linux-amd64
                  +##portable shell syntax
                  +
                  +[ -n "${GOROOT}" ] && export GOROOT=${GOROOT}
                  +[ -n "/usr/local/go1.10.1.linux-amd64" ] && export GOROOT=/usr/local/go1.10.1.linux-amd64
                  +
                  +[ -n "${GOPATH}" ] && export GOPATH=${GOPATH}
                  +[ -n "/usr/local/golang/packages" ] && export GOPATH=/usr/local/golang/packages
                  +
                  +export PATH=$PATH:${GOROOT}/go/bin
                  +export PATH=${PATH}:/usr/local/go1.10.1.linux-amd64/go/bin
----------
          ID: golang-package-archive-install-home-alternative-install
    Function: cmd.run
        Name: update-alternatives --install /usr/local/go golang-home /usr/local/go1.10.1.linux-amd64/go 1000
      Result: True
     Comment: onlyif condition is false
     Started: 17:50:52.505840
    Duration: 0.429 ms
     Changes:   
----------
          ID: golang-package-archive-install-home-alternative-set
    Function: alternatives.set
        Name: golang-home
      Result: True
     Comment: Alternative for golang-home already set to /usr/local/go1.10.1.linux-amd64/go
     Started: 17:50:52.506404
    Duration: 4.273 ms
     Changes:   
----------
          ID: golang-package-archive-install-go-alternative-install
    Function: cmd.run
        Name: update-alternatives --install /usr/bin/go link-go /usr/local/go1.10.1.linux-amd64/go/bin/go 1000
      Result: True
     Comment: onlyif condition is false
     Started: 17:50:52.510814
    Duration: 0.431 ms
     Changes:   
----------
          ID: golang-package-archive-install-go-alternative-set
    Function: alternatives.set
        Name: link-go
      Result: True
     Comment: Alternative for link-go already set to /usr/local/go1.10.1.linux-amd64/go/bin/go
     Started: 17:50:52.511593
    Duration: 4.075 ms
     Changes:   
----------
          ID: golang-package-archive-install-godoc-alternative-install
    Function: cmd.run
        Name: update-alternatives --install /usr/bin/godoc link-godoc /usr/local/go1.10.1.linux-amd64/go/bin/godoc 1000
      Result: True
     Comment: onlyif condition is false
     Started: 17:50:52.515809
    Duration: 0.372 ms
     Changes:   
----------
          ID: golang-package-archive-install-godoc-alternative-set
    Function: alternatives.set
        Name: link-godoc
      Result: True
     Comment: Alternative for link-godoc already set to /usr/local/go1.10.1.linux-amd64/go/bin/godoc
     Started: 17:50:52.516469
    Duration: 4.187 ms
     Changes:   
----------
          ID: golang-package-archive-install-gofmt-alternative-install
    Function: cmd.run
        Name: update-alternatives --install /usr/bin/gofmt link-gofmt /usr/local/go1.10.1.linux-amd64/go/bin/gofmt 1000
      Result: True
     Comment: onlyif condition is false
     Started: 17:50:52.520795
    Duration: 0.368 ms
     Changes:   
----------
          ID: golang-package-archive-install-gofmt-alternative-set
    Function: alternatives.set
        Name: link-gofmt
      Result: True
     Comment: Alternative for link-gofmt already set to /usr/local/go1.10.1.linux-amd64/go/bin/gofmt
     Started: 17:50:52.521450
    Duration: 4.147 ms
     Changes:   

Summary for local
-------------
Succeeded: 16 (changed=2)
Failed:     0
-------------
Total states run:     16



vagrant@ubuntu1804:~$ go env GOPATH
/home/vagrant/go
vagrant@ubuntu1804:~$ go env GOROOT
/usr/local/go
vagrant@ubuntu1804:~$ which go
/usr/bin/go

```

**Comment by noelmcloughlin** [comment]
Fixed integration tests.  BTW, this PR has no breaking changes.

**Comment by noelmcloughlin** [comment]
Fixed comment syntax so lint passes.

Regardless of this PR I am concerned that GOPATH is exported incorrectly in this formula.
<moved text to https://github.com/saltstack-formulas/golang-formula/issues/28 >

Hi @jackscott , could you help.

**Comment by noelmcloughlin** [comment]
See https://golang.org/doc/code.html#GOPATH and https://github.com/golang/go/wiki/SettingGOPATH

**Comment by noelmcloughlin** [comment]
To unblock this PR I raised https://github.com/saltstack-formulas/golang-formula/issues/28

**Comment by myii** [comment]
Sorry for the delay, @noelmcloughlin.  There looks like quite a few changes here, so my plan is to run this through a little process I have locally, to compare before and after.  Once I have the results, I'll share them with you here.

**Comment by myii** [comment]
@noelmcloughlin Just to add, this may take me a day or two but I'll definitely come back to this (bookmarked).  Please feel free to ping me if I take too long.

**Comment by myii** [comment]
@noelmcloughlin I've got some results that I'll be sharing with you soon.  I haven't forgotten about you!

**Comment by myii** [comment]
@noelmcloughlin It only took me another 18 days!  Sorry about that.  So here it is:

* https://github.com/myii/golang-formula/commit/984a74c625cb86d92b59378db7a50f17e03fa9fc

If you can look through that diff and confirm that's the exact map changes you were expecting, then we're a step closer to getting this merged.  Just to be clear, I've run a `yaml_dump` state locally, that extracts the map during the Travis run, for each instance.  I did that before and after and that gives us a diff of what has changed in the proposed PR.  What do you think about having such a feature used in our formulas?

**Comment by noelmcloughlin** [comment]
Hi @myii I have looked through the diff and the changes are as expected. 

This yaml_dump_diff feature could be useful. 
It's hard to judge in this case because osfamilymap.yaml is empty for Linux OSes.

**Comment by myii** [comment]
> Hi @myii I have looked through the diff and the changes are as expected.

@noelmcloughlin Great.  I'd be more comfortable with a second set of eyes to look through the changes.  Any suggestions on who to ping?

**Comment by noelmcloughlin** [comment]
@jackscott I guess.

**Comment by noelmcloughlin** [comment]
@myii can we merge this - no second pair of eyes is available ;-)

**Comment by myii** [comment]
@noelmcloughlin Just a brief point.  We don't need to include these files any more since they have been centralised into the https://github.com/saltstack-formulas/.github repo:

* `.github/ISSUE_TEMPLATE/bug_report.md`
* `.github/ISSUE_TEMPLATE/feature_request.md`

**Comment by noelmcloughlin** [comment]
selfie merging as this is open a long time without further comments.
I squashed some commits and fixed rubocop problems. 
Any issues can be addressed in followup PR.

File: Gemfile [hunk 1]
```diff
@@ -5,7 +5,3 @@ source 'https://rubygems.org'
 gem 'kitchen-docker', '>= 2.9'
 gem 'kitchen-inspec', '>= 1.1'
 gem 'kitchen-salt', '>= 0.6.0'
-# Latest versions of `train` cause failure when running `kitchen verify`
-# Downgrading to `3.2.0` until this is fixed upstream
-# https://github.com/inspec/train/pull/544#issuecomment-566055052
-gem 'train', '3.2.0'
```
Comment (myii) [inline Gemfile:11]: @noelmcloughlin Why did you remove this section?  The bug hasn't been resolved yet:

* https://github.com/inspec/train/issues/548#issuecomment-577990651

Comment (noelmcloughlin) [reply Gemfile:11]: I inadvertently dropped trying to merge masters and squash commits. I'll raise a PR to restore this now.

Comment (noelmcloughlin) [reply Gemfile:11]: https://github.com/saltstack-formulas/golang-formula/pull/34

Comment (myii) [reply Gemfile:11]: OK, thanks.
